### PR TITLE
o/hookstate/ctlcmd: queue service commands if run from default-configure hook

### PR DESCRIPTION
**Pull request by ZeyadYasser**
This PR queues `snapctl` services' commands to be run after the `default-configure` similar to what is done for configure hook #4070. This is to avoid a problem where the service doesn't see a new value if it uses `snapctl get ...`, because it's still not committed by `default-configure` hook similar to what was described in [LP#2047949](https://bugs.launchpad.net/snapd/+bug/2047949) .

Fixes: https://bugs.launchpad.net/snapd/+bug/2047949

### CODE REVIEW

**Review by zyga — DISMISSED** [review]
Looks good.

Please make sure to link updated documentation on the snapcraft forum.

**Comment by ZeyadYasser** [comment]
> Please make sure to link updated documentation on the snapcraft forum.

This fix just makes the implementation consistent with the documented behavior https://snapcraft.io/docs/supported-snap-hooks#heading--default-configure and the snapctl docs don't mention this behavior at all.

Do you have a specific doc page in mind that would need updating?

**Comment by ernestl** [comment]
It's not clear why the user still wants to start services from within the default configure hook. My understanding is the the reason for this was specific to the order of configure-hook VS starting services as part of the do-install flow which should not be required anymore unless I am missing something. 

A secondary more subtle point, it does not seem like a good idea to start/interfere with service commands before do-install start service is done. Longer term we might want explicitly limit snapctl powers depending on how far the install sequence progressed to avoid creative permutations that cause problems.

**Comment by ernestl** [comment]
> It's not clear why the user still wants to start services from within the default configure hook. My understanding is the the reason for this was specific to the order of configure-hook VS starting services as part of the do-install flow which should not be required anymore unless I am missing something.
> 
> A secondary more subtle point, it does not seem like a good idea to start/interfere with service commands before do-install start service is done. Longer term we might want explicitly limit snapctl powers depending on how far the install sequence progressed to avoid creative permutations that cause problems.

The reason why the user have service disabled still: They use the snap for core and for classic, and in classic there is no concept of seeding configuration with gadget. They rely on the user manually modifying the configuration and starting the service, so it is required to have it disabled by default.

File: overlord/hookstate/ctlcmd/services_test.go [hunk 1]
```diff
@@ -481,10 +481,11 @@ func (s *servicectlSuite) TestQueuedCommands(c *C) {
...
 	for _, ts := range tts {
 		tsTasks := ts.Tasks()
-		// assumes configure task is last
-		task := tsTasks[len(tsTasks)-1]
+
+		// default-configure hook runs before start-snap-services which runs before the configure hook
+		task := tsTasks[len(tsTasks)-4]
 		c.Assert(task.Kind(), Equals, "run-hook")
-		setup := &hookstate.HookSetup{Snap: "test-snap", Revision: snap.R(1), Hook: "configure"}
+		setup := &hookstate.HookSetup{Snap: "test-snap", Revision: snap.R(1), Hook: "default-configure"}
```
Comment (pedronis) [inline overlord/hookstate/ctlcmd/services_test.go:488]: it's strange to change preexisting tests instead of adding new ones, you can also write a shared helper that takes the hook and does the shared part between the two kind of tests

File: overlord/hookstate/ctlcmd/helpers.go [hunk 2]
```diff
@@ -149,7 +149,7 @@ func runServiceCommand(context *hookstate.Context, inst *servicestate.Instructio
 		return err
 	}
 
-	if !context.IsEphemeral() && context.HookName() == "configure" {
+	if !context.IsEphemeral() && (context.HookName() == "configure" || context.HookName() == "default-configure") {
```
Comment (pedronis) [inline overlord/hookstate/ctlcmd/helpers.go:152]: where do the tasks queued after default-configure exactly end up? it's not very transparent from tests

**Review by pedronis — COMMENTED** [review]
some question/comments

File: overlord/hookstate/ctlcmd/helpers.go (hunk 2 above)
Comment (ZeyadYasser) [reply overlord/hookstate/ctlcmd/helpers.go:152]: They queued at the end after all install tasks
https://github.com/snapcore/snapd/blob/0bf3e39e00948e76ed45cbcc96facee8da181395/overlord/hookstate/ctlcmd/helpers.go#L114-L124

File: overlord/hookstate/ctlcmd/services_test.go (hunk 1 above)
Comment (ZeyadYasser) [reply overlord/hookstate/ctlcmd/services_test.go]: Thanks for pointing this out, I refactored the test into two test with a shared helper

File: overlord/hookstate/ctlcmd/helpers.go (hunk 2 above)
Comment (ZeyadYasser) [reply overlord/hookstate/ctlcmd/helpers.go:152]: I added clarifying comments in the test

Comment (pedronis) [reply overlord/hookstate/ctlcmd/helpers.go:152]: does this mean after the configure hook? does this solve the original issue people were having in ways they expect?

File: overlord/hookstate/ctlcmd/services_test.go (hunk 1 above)
Comment (pedronis) [reply overlord/hookstate/ctlcmd/services_test.go]: thanks for the refactor

File: overlord/hookstate/ctlcmd/helpers.go (hunk 2 above)
Comment (ZeyadYasser) [reply overlord/hookstate/ctlcmd/helpers.go:152]: Yes, after both the configure hook and the health-check hook.

For illustration this is the expected order of tasks for installing a snap which is used in the unit tests:
```
"prerequisites",
"download-snap",
"validate-snap",
"mount-snap",
"copy-snap-data",
"setup-profiles",
"link-snap",
"auto-connect",
"set-auto-aliases",
"setup-aliases",
"run-hook[install]",
"run-hook[default-configure]",
"start-snap-services",
"run-hook[configure]",
"run-hook[check-health]",
"exec-command",                      <----  task for snapctl command
"service-control",                       <----  task for service operation
```

Comment (pedronis) [reply overlord/hookstate/ctlcmd/helpers.go:152]: I see, we probably want to ask the reporter on the bugs if this is too late for them or ok?

Comment (ZeyadYasser) [reply overlord/hookstate/ctlcmd/helpers.go:152]: > does this solve the original issue people were having in ways they expect?

Yes, I tested it locally using the snap they provided. I left a comment there but I am waiting for feedback.

Comment (ZeyadYasser) [reply overlord/hookstate/ctlcmd/helpers.go:152]: > I see, we probably want to ask the reporter on the bugs if this is too late for them or ok?

Good point, I will check.

**Review by pedronis — APPROVED** [review]
this is fine with me assuming it solves the original problems as is

File: overlord/hookstate/ctlcmd/helpers.go (hunk 2 above)
Comment (farshidtz) [reply overlord/hookstate/ctlcmd/helpers.go:152]: Thanks @ZeyadYasser for the fix and your explanations.

The use case is from a snap that has a one-shot service doing several firewall and network configuration to set up a Thread bridge (linked in the bug report). As mentioned in https://github.com/snapcore/snapd/pull/13960#issuecomment-2133693371, the services are disabled by default, so that the user (on classic but also a pre-built Core image) gets a chance to configure for e.g. the networking interface before starting the services. However, the same snap needs to start automatically on Core when the config is seeded from a gadget. (A workaround was suggested by @zyga to keep services enabled by default but not run them when the needed config isn't set; this could work well for a limited config set.)

In our case, with the workflow proposed in this PR, it could work as follows:
1. `default-configure` hook calls `snapctl start` if the `autostart` config option is set to true.
2. `configure` hook does some validation (not done in `default-configure` to avoid redundant checks).
3. services start.

But I suppose this isn't valid in every use case, as some applications expect the `configure` hook to execute while the services are running.

In summary, while the proposed order isn't an issue for our use case, I believe starting the services right after the hook succeeds makes this more inline with the snap best practices.

Side note: I wonder why `run-hook[check-health]` is ordered between `configure` and `exec-command`/`service-control` tasks.

Another solution I can think of is to add the ability to just enable/disable services via `snapctl` without starting or stopping them. Then, the `configure-hook` could simply call for e.g. the `snapctl enable --service <service>` (`snapctl enable` does something else) command (without needing to queue) and the following task (`start-snap-services`) should just take care of the startup. I've already discussed this with @ernestl, and I'm not sure if this is going to work. I have experimented by calling `snapctl stop --disable <service>` from `install` and `default-configure` hooks for an **enabled** service, and it works as expected: the service gets disabled and never starts.

Comment (pedronis) [reply overlord/hookstate/ctlcmd/helpers.go:152]: the earliest we can start those services is after snap-start-services,  @ZeyadYasser could you look what would be the complexity of queuing commands from default-configure there?

Comment (ZeyadYasser) [reply overlord/hookstate/ctlcmd/helpers.go:152]: We could have a variant of `queueCommand` or some special handling code that does the following:
- if the service commands are coming from `default-configure` hook
- simply inject the commands' tasksets after `start-snap-services` using `snapstate.InjectTasks`
    - `snapstate.InjectTasks` takes care of the complexity of adjusting the tasks' dependencies

Comment (ZeyadYasser) [reply overlord/hookstate/ctlcmd/helpers.go]: The PR is now updated to inject service commands after `start-snap-services` for the `default-configure` hook.

**Review by bboozzoo — APPROVED** [review]
LGTM

File: overlord/hookstate/ctlcmd/helpers.go [hunk 3]
```diff
@@ -122,6 +136,38 @@ func queueCommand(context *hookstate.Context, tts []*state.TaskSet) error {
...
+	st.Lock()
+	defer st.Unlock()
+
+	_, tasks, err := prepareQueueCommand(context, tts)
+	if err != nil {
+		return err
+	}
+
+	for _, t := range tasks {
+		if t.Kind() == "start-snap-services" {
```
Comment (pedronis) [inline overlord/hookstate/ctlcmd/helpers.go:163]: I fear that this a bit too simple for the case where the change is using a single lane/transaction=all-snaps, as there might be multiple start-snap-services involved, as there might just one lane for everything. I think this needs to check also that the snap-name of this task matches the snap name of the hook at least,  we probably need a test somewhere for this or a spread test.  configure has probably the same problem already but there is less noticeable because it would end waiting for all the snaps in the change, which maybe we want to fix as well

**Review by pedronis — CHANGES_REQUESTED** [review]
I think there are issues related to where to inject if lanes are shared like if transaction=all-snaps is used

File: overlord/hookstate/ctlcmd/helpers.go (hunk 3 above)
Comment (ZeyadYasser) [reply overlord/hookstate/ctlcmd/helpers.go:168]: Great catch, Thank you!

I added a test to trigger this edge case and fixed this bug but only for default-configure hook since configure hook is not affected that much and have been like that since the start. The reason I skipped the fix for configure hook is because there is no straightforward way to associate an arbitrary task kind with a snap name.

File: overlord/hookstate/ctlcmd/services_test.go [hunk 4]
```diff
@@ -512,7 +513,183 @@ func (s *servicectlSuite) TestQueuedCommands(c *C) {
...
+
+func (s *servicectlSuite) TestQueuedCommandsOrderingDefaultConfigureHookSingleTransaction(c *C) {
+	const hook = "default-configure"
+	const singleTransaction = true
+	s.testQueuedCommandsOrdering(c, hook, singleTransaction)
+}
+
+func (s *servicectlSuite) TestQueuedCommandsOrderingConfigureHook(c *C) {
+	const hook = "configure"
+	const singleTransaction = false
```
Comment (pedronis) [inline overlord/hookstate/ctlcmd/services_test.go:688]: don't we need the variant when this is true? or is the check above not working for that case?  if we want not to test we should at least leave a comment somewhere

**Review by pedronis — COMMENTED** [review]
thank you, test question

File: overlord/hookstate/ctlcmd/services_test.go (hunk 4 above)
Comment (ZeyadYasser) [reply overlord/hookstate/ctlcmd/services_test.go:688]: Yes, It doesn't work for configure hook. It is tricky to get snap name for all task kinds in the case of configure hook so I didn't add the single transaction fix for configure hook since it is a non-issue due how commands are queued at the very end .

I added the test (but commented out) with a comment explaining why it is left untested.

**Review by pedronis — APPROVED** [review]
thanks

**Review by bboozzoo — APPROVED** [review]
Thanks

**Comment by ZeyadYasser** [comment]
Rebased to fix failing arch tests
